File: opentt/frontend/src/services/institutionService.ts
import api from '@/lib/api';
import type {
  Institution,
  InstitutionCreate,
  InstitutionUpdate,
  AcademicYear,
  AcademicYearCreate,
  DayOfWeek,
  DayOfWeekCreate,
  DayOfWeekUpdate,
  Period,
  PeriodCreate,
  PeriodUpdate,
  PaginationParams,
} from '@/types';

// Institutions
export const institutionService = {
  getAll: async (params?: PaginationParams) => {
    const response = await api.get<Institution[]>('/api/institutions/', { params });
    return response.data;
  },

  getById: async (id: number) => {
    const response = await api.get<Institution>(`/api/institutions/${id}`);
    return response.data;
  },

  create: async (data: InstitutionCreate) => {
    const response = await api.post<Institution>('/api/institutions/', data);
    return response.data;
  },

  update: async (id: number, data: InstitutionUpdate) => {
    const response = await api.put<Institution>(`/api/institutions/${id}`, data);
    return response.data;
  },

  delete: async (id: number) => {
    await api.delete(`/api/institutions/${id}`);
  },
};

// Academic Years
export const academicYearService = {
  getAll: async (institutionId: number, params?: PaginationParams) => {
    const response = await api.get<AcademicYear[]>(
      `/api/institutions/${institutionId}/academic-years`,
      { params }
    );
    return response.data;
  },

  create: async (institutionId: number, data: AcademicYearCreate) => {
    const response = await api.post<AcademicYear>(
      `/api/institutions/${institutionId}/academic-years`,
      data
    );
    return response.data;
  },

  update: async (institutionId: number, yearId: number, data: Partial<AcademicYearCreate>) => {
    const response = await api.put<AcademicYear>(
      `/api/institutions/${institutionId}/academic-years/${yearId}`,
      data
    );
    return response.data;
  },

  delete: async (institutionId: number, yearId: number) => {
    await api.delete(`/api/institutions/${institutionId}/academic-years/${yearId}`);
  },
};

// Days of Week
export const dayOfWeekService = {
  getAll: async (institutionId: number) => {
    const response = await api.get<DayOfWeek[]>(`/api/institutions/${institutionId}/days`);
    return response.data;
  },

  create: async (institutionId: number, data: DayOfWeekCreate) => {
    const response = await api.post<DayOfWeek>(
      `/api/institutions/${institutionId}/days`,
      data
    );
    return response.data;
  },

  update: async (institutionId: number, dayId: number, data: DayOfWeekUpdate) => {
    const response = await api.put<DayOfWeek>(
      `/api/institutions/${institutionId}/days/${dayId}`,
      data
    );
    return response.data;
  },

  delete: async (institutionId: number, dayId: number) => {
    await api.delete(`/api/institutions/${institutionId}/days/${dayId}`);
  },
};

// Periods
export const periodService = {
  getAll: async (institutionId: number) => {
    const response = await api.get<Period[]>(`/api/institutions/${institutionId}/periods`);
    return response.data;
  },

  create: async (institutionId: number, data: PeriodCreate) => {
    const response = await api.post<Period>(
      `/api/institutions/${institutionId}/periods`,
      data
    );
    return response.data;
  },

  update: async (institutionId: number, periodId: number, data: PeriodUpdate) => {
    const response = await api.put<Period>(
      `/api/institutions/${institutionId}/periods/${periodId}`,
      data
    );
    return response.data;
  },

  delete: async (institutionId: number, periodId: number) => {
    await api.delete(`/api/institutions/${institutionId}/periods/${periodId}`);
  },
};

File: opentt/backend/app/services/institution_service.py
from typing import Optional
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from sqlalchemy.orm import selectinload

from app.models.institution import Institution, AcademicYear, DayOfWeek, Period
from app.schemas.institution import (
    InstitutionCreate,
    InstitutionUpdate,
    AcademicYearCreate,
    AcademicYearUpdate,
    DayOfWeekCreate,
    DayOfWeekUpdate,
    PeriodCreate,
    PeriodUpdate,
)
from app.services.base import BaseService


class InstitutionService(BaseService[Institution, InstitutionCreate, InstitutionUpdate]):
    def __init__(self):
        super().__init__(Institution)

    async def get_by_slug(self, db: AsyncSession, slug: str) -> Optional[Institution]:
        """Get institution by slug"""
        result = await db.execute(select(Institution).where(Institution.slug == slug))
        return result.scalar_one_or_none()

    async def get_with_setup(self, db: AsyncSession, institution_id: int) -> Optional[Institution]:
        """Get institution with days, periods, and academic years"""
        result = await db.execute(
            select(Institution)
            .where(Institution.id == institution_id)
            .options(
                selectinload(Institution.days_of_week),
                selectinload(Institution.periods),
                selectinload(Institution.academic_years),
            )
        )
        return result.scalar_one_or_none()


class AcademicYearService(BaseService[AcademicYear, AcademicYearCreate, AcademicYearUpdate]):
    def __init__(self):
        super().__init__(AcademicYear)


class DayOfWeekService(BaseService[DayOfWeek, DayOfWeekCreate, DayOfWeekUpdate]):
    def __init__(self):
        super().__init__(DayOfWeek)

    async def get_by_institution(
        self, db: AsyncSession, institution_id: int, active_only: bool = True
    ):
        """Get all days for an institution, optionally filtered by active status"""
        from sqlalchemy import and_
        
        query = select(DayOfWeek).where(DayOfWeek.institution_id == institution_id)
        if active_only:
            query = query.where(DayOfWeek.is_active == True)
        query = query.order_by(DayOfWeek.order)
        
        result = await db.execute(query)
        return list(result.scalars().all())


class PeriodService(BaseService[Period, PeriodCreate, PeriodUpdate]):
    def __init__(self):
        super().__init__(Period)

    async def get_by_institution(self, db: AsyncSession, institution_id: int):
        """Get all periods for an institution ordered by order"""
        query = (
            select(Period)
            .where(Period.institution_id == institution_id)
            .order_by(Period.order)
        )
        result = await db.execute(query)
        return list(result.scalars().all())


# Service instances
institution_service = InstitutionService()
academic_year_service = AcademicYearService()
day_of_week_service = DayOfWeekService()
period_service = PeriodService()
